File: go/201_bitwise_and_of_numbers_range.go
func rangeBitwiseAnd(m int, n int) int {
	power_of_2 := 1
	for power_of_2 <= m {
		power_of_2 *= 2
	}
	if power_of_2 < n {
		return 0
	}

	result := m
	for i := m; i <= n; i++ {
		result &= i
	}
	return result
}

File: ruby/201_bitwise_and_of_numbers_range.rb

# Faster than 100.00% of submissions
# @param {Integer} m
# @param {Integer} n
# @return {Integer}
def range_bitwise_and(m, n)
  # If n is a "place" larger than m (i.e., if there is a power of 2 between m and n) then there will be no overlap
  power_of_2 = 1
  power_of_2 *= 2 while power_of_2 <= m
  return 0 if power_of_2 <= n

  result = m
  (m+1..n).each do |num|
    result &= num
  end
  result
end
